#### Análise de Dados com PYTHON.
- Agrupando Valores


#### * Carregando dados com PANDAS

In [2]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas

###### A fim de evitar ficar chamando a todo momento uma biblioteca podemos atribuir a elas, apelidos, para mais rapidez e facilidade nos códigos
###### Porém estes apelidos seguem padrões dentro da comunidade PY, exemplo disso é a pandas que tem como apelido o pd

In [4]:
import pandas as pd

In [5]:
# importando dados de um arquivo csv - lembrando que o arquivo deve estar dentro da nossa pasta de desenvolvimento

dados = pd.read_csv("base_dados.csv")                    # read = ler

In [6]:
# verificando as primeiras linhas carregadas

dados.head()                                             # head = cabeçalho

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
0             SP         139.5  2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4  2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5  2019-02-14   Produto Errado  No Prazo  
3             SP         326.0  2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6  2019-03-07       S/ Devolu.  Atrasado

In [7]:
dados.head(10)                        # verificando as 10 primeiras linhas

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   
5           A16         WALMART       JOÃO GOMES        2019-02-23   
6           A17         WALMART       JOÃO GOMES        2019-02-24   
7           A20  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-26   
8           A22      AMERICANAS  VALDIR ESPINOSA        2019-03-01   
9           A23      AMERICANAS     MARCOS LEROY        2019-03-02   

  estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
0             SP         139.5  2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4  2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5  2019-02-14   Produto Errado  No Prazo  
3             SP         326.0  2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6  2019-03-07       S/ Devolu.  Atrasado  
5             RJ         467.6  2019-03-06   Produto Errado  Atrasado  
6             SP         297.0  2019-03-09   Produto Errado  Atrasado  
7             BH         473.2  2019-03-07   Produto Errado  No Prazo  
8             SP         165.0  2019-03-13       Danificado  Atrasado  
9             RJ          31.0  2019-03-06       Danificado  No Prazo

In [8]:
# verificar as últimas linhas carregadas

dados.tail()                                     # tail = rabo (últimas)

numero_pedido         cliente        motorista dt_emissao_pedido  \
4277         A9675  MAGAZINE LUIZA  VALDIR ESPINOSA        2021-11-27   
4278         A9676         WALMART      LUIZ PARDAL        2021-11-28   
4279         A9680    CASA E VIDEO   TÚLIO SILVEIRA        2021-12-02   
4280         A9681  RICARDO ELETRO     MARCOS LEROY        2021-12-03   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
4277             RJ         763.2  2021-12-15   Produto Errado  Atrasado  
4278             SP         226.8  2021-12-12   Produto Errado  Atrasado  
4279             SP         821.1  2021-12-20       Danificado  Atrasado  
4280             SP          49.8  2021-12-17       S/ Devolu.  Atrasado  
4281             BH         715.5  2021-12-17       Danificado  Atrasado

In [9]:
# verificando o tamanho da base de dados, que será apresentado pelo comando shape, a quantidade de linhas e de colunas

dados.shape                           # por ser uma propriedade e não um método o shape não exige os parênteses

(4282, 9)

#### Carregando dados com Excel

###### - Recomenda-se, por uma boa prática, que se faça uma análise exploratória dos dados recebidos antes de joga-lo em nosso sistema.
###### - Mas do que se trata essa análise exploratória?
######   * Abrir o arquivo recebido, saber sobre suas abas e se não há abas ocultas, se houver deve-se abri-las;
######   * Estudar e entender os dados e as colunas do arquivo recebido;
######   * Saber o número de linhas existentes nesses arquivos, no caso do excell para chegar nessa informação -> na linha A2 CTRL + SHIFT + seta p/ baixo
       

In [10]:
dados = pd.read_excel("base_dados.xlsx")

In [11]:
# Verificando as primeiras linhas

dados.head()


numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [12]:
# Verificando as últimas linhas

dados.tail()


numero_pedido         cliente        motorista dt_emissao_pedido  \
4277         A9675  MAGAZINE LUIZA  VALDIR ESPINOSA        2021-11-27   
4278         A9676         WALMART      LUIZ PARDAL        2021-11-28   
4279         A9680    CASA E VIDEO   TÚLIO SILVEIRA        2021-12-02   
4280         A9681  RICARDO ELETRO     MARCOS LEROY        2021-12-03   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
4277             RJ         763.2 2021-12-15   Produto Errado  Atrasado  
4278             SP         226.8 2021-12-12   Produto Errado  Atrasado  
4279             SP         821.1 2021-12-20       Danificado  Atrasado  
4280             SP          49.8 2021-12-17       S/ Devolu.  Atrasado  
4281             BH         715.5 2021-12-17       Danificado  Atrasado

In [13]:
# Verificando o tamanho da base de dados

dados.shape                  # Lembrando que,shape, por ser uma propriedade e não um método não exige os parêntese no final

(4282, 9)

#### Verificando o tipo de dados das colunas

In [14]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4282 entries, 0 to 4281
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   numero_pedido      4282 non-null   object        
 1   cliente            4282 non-null   object        
 2   motorista          4282 non-null   object        
 3   dt_emissao_pedido  4282 non-null   datetime64[ns]
 4   estado_destino     4282 non-null   object        
 5   valor_pedido       4282 non-null   float64       
 6   dt_entrega         4282 non-null   datetime64[ns]
 7   motivo_devolucao   4282 non-null   object        
 8   status             4282 non-null   object        
dtypes: datetime64[ns](2), float64(1), object(6)
memory usage: 301.2+ KB


     - As informações acima nos trás: 
- 4282 linhas no arquivo; 
- em 9 colunas; 
- o nome das colunas; 
- non-null, que indica que não há nenhuma linha sem dado; (no-null deve ser aprofundado, pois trat-se de uma análise mais apurada)
- e o dtype, que trás a informação de 'object', que indica que é uma strings
- datetime que é uma coluna de data e hora;
- há na tabela valores float, informações não inteiras (float64)

In [15]:
# Podemos verificar o tipo de dado da variável dados

type(dados)

pandas.core.frame.DataFrame

#### Acessando Colunas individualmente
- Dados tabular -> trata-se de uma tabela com uma sequência de linhas e colunas.

In [16]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [17]:
# Acessando a coluna cliente

dados["cliente"]

0              WALMART
1       MAGAZINE LUIZA
2       MAGAZINE LUIZA
3       MAGAZINE LUIZA
4              WALMART
             ...      
4277    MAGAZINE LUIZA
4278           WALMART
4279      CASA E VIDEO
4280    RICARDO ELETRO
4281        AMERICANAS
Name: cliente, Length: 4282, dtype: object

In [18]:
# Acessando a coluna número pedido

dados["numero_pedido"]

0          A3
1          A4
2          A7
3         A12
4         A15
        ...  
4277    A9675
4278    A9676
4279    A9680
4280    A9681
4281    A9683
Name: numero_pedido, Length: 4282, dtype: object

In [19]:
# Acessando multiplas colunas

dados[["cliente", "estado_destino"]]

cliente estado_destino
0            WALMART             SP
1     MAGAZINE LUIZA             RJ
2     MAGAZINE LUIZA             SP
3     MAGAZINE LUIZA             SP
4            WALMART             RJ
...              ...            ...
4277  MAGAZINE LUIZA             RJ
4278         WALMART             SP
4279    CASA E VIDEO             SP
4280  RICARDO ELETRO             SP
4281      AMERICANAS             BH

[4282 rows x 2 columns]

In [20]:
# Dentro da análise podemos criar uma lista recebendo o nome de uma varíavel e solicita-la no Data Frame.

colunas = ["cliente", "motorista", "estado_destino","status"]

dados[colunas]

cliente        motorista estado_destino    status
0            WALMART     FELIPE SILVA             SP  No Prazo
1     MAGAZINE LUIZA   TÚLIO SILVEIRA             RJ  Atrasado
2     MAGAZINE LUIZA  VALDIR ESPINOSA             SP  No Prazo
3     MAGAZINE LUIZA     MARCOS LEROY             SP  No Prazo
4            WALMART      LUIZ PARDAL             RJ  Atrasado
...              ...              ...            ...       ...
4277  MAGAZINE LUIZA  VALDIR ESPINOSA             RJ  Atrasado
4278         WALMART      LUIZ PARDAL             SP  Atrasado
4279    CASA E VIDEO   TÚLIO SILVEIRA             SP  Atrasado
4280  RICARDO ELETRO     MARCOS LEROY             SP  Atrasado
4281      AMERICANAS     FELIPE SILVA             BH  Atrasado

[4282 rows x 4 columns]

#### Acessando as Linhas

In [21]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

###### Os números sequênciais nas linhas, são índices criados automaticamente pelo carregamento do sistema e são idependentes das sequências da página do excel.
###### PANDAS trata-se de uma biblioteca muito poderosa e entre esses poderes há várias formas de se chamar dados. Porém como estamos no aprendizado de melhores práticas, vamos fazer a seleção de linhas pelo método loc (localization).

In [22]:
dados.loc[0]

numero_pedido                         A3
cliente                          WALMART
motorista                   FELIPE SILVA
dt_emissao_pedido    2019-01-27 00:00:00
estado_destino                        SP
valor_pedido                       139.5
dt_entrega           2019-02-03 00:00:00
motivo_devolucao              S/ Devolu.
status                          No Prazo
Name: 0, dtype: object

In [23]:
dados.loc[77]

numero_pedido                       A185
cliente                       AMERICANAS
motorista                    LUIZ PARDAL
dt_emissao_pedido    2019-07-05 00:00:00
estado_destino                        BH
valor_pedido                       806.4
dt_entrega           2019-07-18 00:00:00
motivo_devolucao          Produto Errado
status                          Atrasado
Name: 77, dtype: object

In [24]:
# Selecionar linhas e colunas com loc

dados.loc[0, "cliente"]

'WALMART'

In [25]:
dados.loc[0, "motorista"]

'FELIPE SILVA'

In [26]:
dados.loc[0, ["cliente", "motorista"]]

cliente           WALMART
motorista    FELIPE SILVA
Name: 0, dtype: object

#### Filtrando Linhas e Colunas
- Utilizando dado Booleano (bool que tem seu status como True or False)

In [27]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [28]:
# Vamos fazer dentro do Data Frame um filtro buscando apenas os dados do WALMART

dados["cliente"] == "WALMART"

0        True
1       False
2       False
3       False
4        True
        ...  
4277    False
4278     True
4279    False
4280    False
4281    False
Name: cliente, Length: 4282, dtype: bool

In [29]:
# Com as informações recebidas acima podemos criar nosso filtro com uma varíavel, a qual pode receber qualquer nome, no caso filtro

filtro = dados["cliente"] == "WALMART"

dados[filtro]

numero_pedido  cliente     motorista dt_emissao_pedido estado_destino  \
0               A3  WALMART  FELIPE SILVA        2019-01-27             SP   
4              A15  WALMART   LUIZ PARDAL        2019-02-21             RJ   
5              A16  WALMART    JOÃO GOMES        2019-02-23             RJ   
6              A17  WALMART    JOÃO GOMES        2019-02-24             SP   
11             A27  WALMART  FELIPE SILVA        2019-03-07             BH   
...            ...      ...           ...               ...            ...   
4257         A9640  WALMART    JOÃO GOMES        2021-11-03             BH   
4268         A9656  WALMART    JOÃO GOMES        2021-11-09             BH   
4270         A9659  WALMART   LUIZ PARDAL        2021-11-10             RJ   
4275         A9672  WALMART    JOÃO GOMES        2021-11-24             SP   
4278         A9676  WALMART   LUIZ PARDAL        2021-11-28             SP   

      valor_pedido dt_entrega motivo_devolucao    status  
0            139.5 2019-02-03       S/ Devolu.  No Prazo  
4            597.6 2019-03-07       S/ Devolu.  Atrasado  
5            467.6 2019-03-06   Produto Errado  Atrasado  
6            297.0 2019-03-09   Produto Errado  Atrasado  
11           907.2 2019-03-21       S/ Devolu.  Atrasado  
...            ...        ...              ...       ...  
4257         534.4 2021-11-20       Danificado  Atrasado  
4268         342.3 2021-11-24       Danificado  Atrasado  
4270          80.5 2021-11-29   Arrependimento  Atrasado  
4275         328.0 2021-12-14   Produto Errado  Atrasado  
4278         226.8 2021-12-12   Produto Errado  Atrasado  

[1157 rows x 9 columns]

In [30]:
# Pórem podemos ter o mesmo resultado usando o loc

dados.loc[filtro, ["cliente", "motorista"]]

cliente     motorista
0     WALMART  FELIPE SILVA
4     WALMART   LUIZ PARDAL
5     WALMART    JOÃO GOMES
6     WALMART    JOÃO GOMES
11    WALMART  FELIPE SILVA
...       ...           ...
4257  WALMART    JOÃO GOMES
4268  WALMART    JOÃO GOMES
4270  WALMART   LUIZ PARDAL
4275  WALMART    JOÃO GOMES
4278  WALMART   LUIZ PARDAL

[1157 rows x 2 columns]

In [31]:
# Se precisar de toda coluna

dados.loc[filtro]

numero_pedido  cliente     motorista dt_emissao_pedido estado_destino  \
0               A3  WALMART  FELIPE SILVA        2019-01-27             SP   
4              A15  WALMART   LUIZ PARDAL        2019-02-21             RJ   
5              A16  WALMART    JOÃO GOMES        2019-02-23             RJ   
6              A17  WALMART    JOÃO GOMES        2019-02-24             SP   
11             A27  WALMART  FELIPE SILVA        2019-03-07             BH   
...            ...      ...           ...               ...            ...   
4257         A9640  WALMART    JOÃO GOMES        2021-11-03             BH   
4268         A9656  WALMART    JOÃO GOMES        2021-11-09             BH   
4270         A9659  WALMART   LUIZ PARDAL        2021-11-10             RJ   
4275         A9672  WALMART    JOÃO GOMES        2021-11-24             SP   
4278         A9676  WALMART   LUIZ PARDAL        2021-11-28             SP   

      valor_pedido dt_entrega motivo_devolucao    status  
0            139.5 2019-02-03       S/ Devolu.  No Prazo  
4            597.6 2019-03-07       S/ Devolu.  Atrasado  
5            467.6 2019-03-06   Produto Errado  Atrasado  
6            297.0 2019-03-09   Produto Errado  Atrasado  
11           907.2 2019-03-21       S/ Devolu.  Atrasado  
...            ...        ...              ...       ...  
4257         534.4 2021-11-20       Danificado  Atrasado  
4268         342.3 2021-11-24       Danificado  Atrasado  
4270          80.5 2021-11-29   Arrependimento  Atrasado  
4275         328.0 2021-12-14   Produto Errado  Atrasado  
4278         226.8 2021-12-12   Produto Errado  Atrasado  

[1157 rows x 9 columns]

###### Lembrando que veio toda Data Frame do filtro WALMART, como foi especificado na váriavel -> filtro = dados["cliente"] == "WALMART"

In [32]:
# Podemos fazer ainda a GRAVAÇÃO dessas informações dentro de um váriavel

dados_walmart = dados.loc[filtro]

dados_walmart

numero_pedido  cliente     motorista dt_emissao_pedido estado_destino  \
0               A3  WALMART  FELIPE SILVA        2019-01-27             SP   
4              A15  WALMART   LUIZ PARDAL        2019-02-21             RJ   
5              A16  WALMART    JOÃO GOMES        2019-02-23             RJ   
6              A17  WALMART    JOÃO GOMES        2019-02-24             SP   
11             A27  WALMART  FELIPE SILVA        2019-03-07             BH   
...            ...      ...           ...               ...            ...   
4257         A9640  WALMART    JOÃO GOMES        2021-11-03             BH   
4268         A9656  WALMART    JOÃO GOMES        2021-11-09             BH   
4270         A9659  WALMART   LUIZ PARDAL        2021-11-10             RJ   
4275         A9672  WALMART    JOÃO GOMES        2021-11-24             SP   
4278         A9676  WALMART   LUIZ PARDAL        2021-11-28             SP   

      valor_pedido dt_entrega motivo_devolucao    status  
0            139.5 2019-02-03       S/ Devolu.  No Prazo  
4            597.6 2019-03-07       S/ Devolu.  Atrasado  
5            467.6 2019-03-06   Produto Errado  Atrasado  
6            297.0 2019-03-09   Produto Errado  Atrasado  
11           907.2 2019-03-21       S/ Devolu.  Atrasado  
...            ...        ...              ...       ...  
4257         534.4 2021-11-20       Danificado  Atrasado  
4268         342.3 2021-11-24       Danificado  Atrasado  
4270          80.5 2021-11-29   Arrependimento  Atrasado  
4275         328.0 2021-12-14   Produto Errado  Atrasado  
4278         226.8 2021-12-12   Produto Errado  Atrasado  

[1157 rows x 9 columns]

In [33]:
# Filtrando todas as linhas onde a coluna estado_destino for igual a BH

filtro_bh = dados["estado_destino"] == "BH"
dados.loc[filtro_bh]

numero_pedido         cliente        motorista dt_emissao_pedido  \
7              A20  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-26   
11             A27         WALMART     FELIPE SILVA        2019-03-07   
14             A35         WALMART     FELIPE SILVA        2019-03-17   
15             A36         WALMART     MARCOS LEROY        2019-03-18   
17             A38         WALMART      LUIZ PARDAL        2019-03-22   
...            ...             ...              ...               ...   
4268         A9656         WALMART       JOÃO GOMES        2021-11-09   
4271         A9662     CASAS BAHIA     FELIPE SILVA        2021-11-12   
4274         A9668    CASA E VIDEO   TÚLIO SILVEIRA        2021-11-15   
4276         A9673     CASAS BAHIA     FELIPE SILVA        2021-11-25   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
7                BH         473.2 2019-03-07   Produto Errado  No Prazo  
11               BH         907.2 2019-03-21       S/ Devolu.  Atrasado  
14               BH         226.8 2019-03-27       Danificado  No Prazo  
15               BH          93.0 2019-04-03   Produto Errado  Atrasado  
17               BH         738.0 2019-03-29       S/ Devolu.  No Prazo  
...             ...           ...        ...              ...       ...  
4268             BH         342.3 2021-11-24       Danificado  Atrasado  
4271             BH         157.0 2021-12-02       S/ Devolu.  Atrasado  
4274             BH         330.0 2021-11-24       Danificado  No Prazo  
4276             BH         907.2 2021-12-15   Produto Errado  Atrasado  
4281             BH         715.5 2021-12-17       Danificado  Atrasado  

[1496 rows x 9 columns]

#### Exportando um DataFrame

In [34]:
dados.loc[filtro_bh]                    # Abaixo temos toda base de dados relacionados com o Estado de Belo Horizonte


numero_pedido         cliente        motorista dt_emissao_pedido  \
7              A20  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-26   
11             A27         WALMART     FELIPE SILVA        2019-03-07   
14             A35         WALMART     FELIPE SILVA        2019-03-17   
15             A36         WALMART     MARCOS LEROY        2019-03-18   
17             A38         WALMART      LUIZ PARDAL        2019-03-22   
...            ...             ...              ...               ...   
4268         A9656         WALMART       JOÃO GOMES        2021-11-09   
4271         A9662     CASAS BAHIA     FELIPE SILVA        2021-11-12   
4274         A9668    CASA E VIDEO   TÚLIO SILVEIRA        2021-11-15   
4276         A9673     CASAS BAHIA     FELIPE SILVA        2021-11-25   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
7                BH         473.2 2019-03-07   Produto Errado  No Prazo  
11               BH         907.2 2019-03-21       S/ Devolu.  Atrasado  
14               BH         226.8 2019-03-27       Danificado  No Prazo  
15               BH          93.0 2019-04-03   Produto Errado  Atrasado  
17               BH         738.0 2019-03-29       S/ Devolu.  No Prazo  
...             ...           ...        ...              ...       ...  
4268             BH         342.3 2021-11-24       Danificado  Atrasado  
4271             BH         157.0 2021-12-02       S/ Devolu.  Atrasado  
4274             BH         330.0 2021-11-24       Danificado  No Prazo  
4276             BH         907.2 2021-12-15   Produto Errado  Atrasado  
4281             BH         715.5 2021-12-17       Danificado  Atrasado  

[1496 rows x 9 columns]

In [35]:
# Entendendo que a análise dos dados está satisfatória e precisamos salvá-la em um arquivo.
# Exportando para o EXCEL

dados.loc[filtro_bh].to_excel("dados_BH.xlsx", index=False)           # dados.loc[filtro_bh] -> solicitado ao PANDAS que pegue esse arquivo
                                                                      # .to_excel            -> exporta essas informações para o EXCEL
                                                                      # "dados_BH.xlsx"      -> com esse nome de arquivo
                                                                      # index=False          -> solicitado ao PANDAS para NÂO exportar o índice(index)

###### se observado nos nossos arquivos no canto superior esquerdo, do VSCode, está criado o arquivo dados_BH.xlsx, bem como na pasta arquivos pessoais do PC.

In [36]:
# Agora vamos Exportar os dados do WALMART, para um arquivo CSV

dados_walmart.to_csv("dados_walmart.csv", index=False)

#### Cálculos com a Base de Dados

#### Gerando estatísticas
- Será gerado estatísticas na coluna valor_pedido
 

In [37]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [38]:
dados['valor_pedido'].describe()

count    4282.000000
mean      339.038230
std       244.817324
min        15.000000
25%       144.900000
50%       279.000000
75%       505.600000
max      1020.000000
Name: valor_pedido, dtype: float64

###### count -> trás os valores de contagem que há na coluna solicitada
###### mean  -> é a média de preço de todos os pedidos
###### std   -> desvio padrão que indica o grau de variação de um conjunto de elementos. (ex.:a média de variação de 3 cálculos)
###### min   -> é o valor de pedido mais barato
###### 25%   -> as porcentagens são conhecidas como QUARTIS e a de 25% é o primeiro quartil, que indica que 25% dos pedidos é até o valor expressado
###### 50%   -> segundo quartil ou mediana e indica que 50% dos pedidos é de até o valor expressado
###### 75%   -> terceiro quartil e indica que 75% dos pedidos é de até o valor expressado
###### max   -> trás o pedido de maior valor

#### Cálculos com PANDAS parte 2 
#### Contagem de Valores Únicos
###### Trata-se de um processo importante pois trás muitas informações importantes dentro de uma análise

In [39]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

##### * Visualizando valores distintos

In [40]:
# Visualizando os valores únicos de uma coluna

dados["cliente"].unique()

array(['WALMART', 'MAGAZINE LUIZA', 'AMERICANAS', 'CASAS BAHIA',
       'CASA E VIDEO', 'RICARDO ELETRO'], dtype=object)

In [41]:
# Visualizando os motoristas da empresa

dados["motorista"].unique()

array(['FELIPE SILVA', 'TÚLIO SILVEIRA', 'VALDIR ESPINOSA',
       'MARCOS LEROY', 'LUIZ PARDAL', 'JOÃO GOMES'], dtype=object)

In [42]:
# Visualizar os estados de destino

dados["estado_destino"].unique()

array(['SP', 'RJ', 'BH'], dtype=object)

#### * Contagem de valores distintos

In [43]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [44]:
# Contando valores distintos para clientes, onde cada loja irá mostrar o número de pedidos

dados["cliente"].value_counts()

cliente
WALMART           1157
CASAS BAHIA       1016
MAGAZINE LUIZA     809
AMERICANAS         591
CASA E VIDEO       385
RICARDO ELETRO     324
Name: count, dtype: int64

In [45]:
# Motorista com o maior número de pedidos

dados["motorista"].value_counts()

motorista
MARCOS LEROY       753
LUIZ PARDAL        750
VALDIR ESPINOSA    732
TÚLIO SILVEIRA     698
JOÃO GOMES         685
FELIPE SILVA       664
Name: count, dtype: int64

In [46]:
# Quantidade de pedidos por estado destino

dados["estado_destino"].value_counts()

estado_destino
BH    1496
SP    1412
RJ    1374
Name: count, dtype: int64

#### Aplicando estatísticas

In [47]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

#### * Média (mean)

In [48]:
dados["valor_pedido"].mean()

# Os resultados de média fazem sentido para colunas numéricas

np.float64(339.0382297991593)

##### *Mediana (median)

In [49]:
dados["valor_pedido"].median()

np.float64(279.0)

##### *Desvio Padrão

In [50]:
dados["valor_pedido"].std()

np.float64(244.81732413883788)

##### *Valor máximo

In [51]:
dados["valor_pedido"].max()

np.float64(1020.0000000000007)

In [52]:
# Qual é a data do último pedido?

dados["dt_emissao_pedido"].max()

Timestamp('2021-12-06 00:00:00')

In [53]:
# Qual é a última data de entrega

dados["dt_entrega"].max()

Timestamp('2021-12-20 00:00:00')

##### *Valor minímo

In [54]:
dados["valor_pedido"].min()

np.float64(15.0)

In [55]:
# Qual é a menor data de pedido?

dados["dt_emissao_pedido"].min()

Timestamp('2019-01-27 00:00:00')

In [56]:
# Qual a entrega mais antiga da base de dados?

dados["dt_emissao_pedido"].min()

Timestamp('2019-01-27 00:00:00')

#### Agrupando Valores

In [57]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

###### Para cada um dos clientes podemos somar os valores dos pedidos, valores de produtos por motoristas,etc

In [59]:
# Mostrando o valor total de pedidos de cada cliente

dados.groupby("cliente")["valor_pedido"].sum()

cliente
AMERICANAS        188903.6
CASA E VIDEO      131108.2
CASAS BAHIA       348189.8
MAGAZINE LUIZA    285890.9
RICARDO ELETRO    104503.3
WALMART           393165.9
Name: valor_pedido, dtype: float64

In [60]:
# Mostrando o valor total de pedidos por motorista

dados.groupby("motorista")["valor_pedido"].sum()

motorista
FELIPE SILVA       224024.7
JOÃO GOMES         237449.6
LUIZ PARDAL        251138.9
MARCOS LEROY       262353.2
TÚLIO SILVEIRA     243329.3
VALDIR ESPINOSA    233466.0
Name: valor_pedido, dtype: float64

In [62]:
# Qual o valor médio de pedido por Estado destino?

dados.groupby("estado_destino")["valor_pedido"].mean()

estado_destino
BH    335.017781
RJ    348.699418
SP    333.896671
Name: valor_pedido, dtype: float64

##### * Agrupando por múltiplas colunas

In [63]:
# Agrupando por cliente e estado destino

dados.groupby(["cliente", "estado_destino",])["valor_pedido"].sum()

cliente         estado_destino
AMERICANAS      BH                 64648.1
                RJ                 66547.8
                SP                 57707.7
CASA E VIDEO    BH                 41805.8
                RJ                 47779.2
                SP                 41523.2
CASAS BAHIA     BH                128301.9
                RJ                110400.7
                SP                109487.2
MAGAZINE LUIZA  BH                 95455.2
                RJ                 94038.0
                SP                 96397.7
RICARDO ELETRO  BH                 30196.0
                RJ                 35755.3
                SP                 38552.0
WALMART         BH                140779.6
                RJ                124592.0
                SP                127794.3
Name: valor_pedido, dtype: float64

In [64]:
# Melhorando a apresentação dos dados acima com o comando to_frame()

dados.groupby(["cliente", "estado_destino",])["valor_pedido"].sum().to_frame()

valor_pedido
cliente        estado_destino              
AMERICANAS     BH                   64648.1
               RJ                   66547.8
               SP                   57707.7
CASA E VIDEO   BH                   41805.8
               RJ                   47779.2
               SP                   41523.2
CASAS BAHIA    BH                  128301.9
               RJ                  110400.7
               SP                  109487.2
MAGAZINE LUIZA BH                   95455.2
               RJ                   94038.0
               SP                   96397.7
RICARDO ELETRO BH                   30196.0
               RJ                   35755.3
               SP                   38552.0
WALMART        BH                  140779.6
               RJ                  124592.0
               SP                  127794.3

In [65]:
# Qual é o total de pedidos por cliente e por estado destino

dados.groupby(["cliente", "estado_destino",])["numero_pedido"].count().to_frame()

numero_pedido
cliente        estado_destino               
AMERICANAS     BH                        205
               RJ                        200
               SP                        186
CASA E VIDEO   BH                        130
               RJ                        132
               SP                        123
CASAS BAHIA    BH                        373
               RJ                        314
               SP                        329
MAGAZINE LUIZA BH                        278
               RJ                        251
               SP                        280
RICARDO ELETRO BH                        101
               RJ                        106
               SP                        117
WALMART        BH                        409
               RJ                        371
               SP                        377

In [70]:
# Qual é o total de pedidos por motorista e por status?

dados.groupby(["motorista", "status"])["numero_pedido"].count().to_frame()

numero_pedido
motorista       status                 
FELIPE SILVA    Atrasado            469
                No Prazo            195
JOÃO GOMES      Atrasado            494
                No Prazo            191
LUIZ PARDAL     Atrasado            570
                No Prazo            180
MARCOS LEROY    Atrasado            567
                No Prazo            186
TÚLIO SILVEIRA  Atrasado            501
                No Prazo            197
VALDIR ESPINOSA Atrasado            534
                No Prazo            198